### PR TITLE
treewide: deprecate pythonForBuild, replace with pythonOnBuildForHost

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes

Deprecate confusing old name, add `lib.warn`.

See

- https://github.com/NixOS/nixpkgs/pull/265570

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/interpreters/python/passthrufun.nix [hunk 1]
```diff
@@ -91,9 +91,9 @@ in rec {
     pythonAtLeast = lib.versionAtLeast pythonVersion;
     pythonOlder = lib.versionOlder pythonVersion;
     inherit hasDistutilsCxxPatch;
-    # TODO: deprecate
-    # Not done immediately because its likely used outside Nixpkgs.
-    pythonForBuild = pythonOnBuildForHost_overridden;
+    pythonForBuild =
+      lib.warn "pythonForBuild is deprecated; please use pythonOnBuildForHost instead"
```
Comment (Artturin) [inline pkgs/development/interpreters/python/passthrufun.nix:95]:
```suggestion
      lib.warnIf (lib.isInOldestRelease 2311) "`pythonForBuild` (from `python*`) has been renamed to `pythonOnBuildForHost`"
```

Same format I used for the `overrideScope'` deprecation in `makeScope`

File: pkgs/development/interpreters/python/passthrufun.nix [hunk 2]
```diff
@@ -91,9 +91,9 @@ in rec {
     pythonAtLeast = lib.versionAtLeast pythonVersion;
     pythonOlder = lib.versionOlder pythonVersion;
     inherit hasDistutilsCxxPatch;
-    # TODO: deprecate
-    # Not done immediately because its likely used outside Nixpkgs.
-    pythonForBuild = pythonOnBuildForHost_overridden;
+    pythonForBuild =
```
Comment (Artturin) [inline pkgs/development/interpreters/python/passthrufun.nix:94]:
```suggestion
    # Remove after 24.11 is released.
    pythonForBuild =
```

3 releases should be enough I think

Comment (Artturin) [reply pkgs/development/interpreters/python/passthrufun.nix:94]: Only used in ~66 places https://github.com/search?q=%2Fpython3.pythonForBuild%2F%20lang%3Anix%20NOT%20is%3Afork&type=code

Comment (amjoseph-nixpkgs) [reply pkgs/development/interpreters/python/passthrufun.nix:95]: > Only used in ~66 places https://github.com/search?q=%2Fpython3.pythonForBuild%2F%20lang%3Anix%20NOT%20is%3Afork&type=code

There are a few more uses that aren't caught by that search string (like `inherit (python3) pythonForBuild`).  This PR should fix all of them.

**Pull request merged**